runtime-modules/forum/src/lib.rs

```diff
@@ -1257,19 +1283,53 @@ impl<T: Trait> Module<T> {
         moderator_id: &T::ModeratorId,
         category_id: &T::CategoryId,
         thread_id: &T::ThreadId,
-    ) -> Result<Thread<T::ForumUserId, T::CategoryId, T::Moment, T::Hash>, &'static str> {
+    ) -> Result<
+        (
+            T::AccountId,
+            Thread<T::ForumUserId, T::CategoryId, T::Moment, T::Hash>,
+        ),
+        &'static str,
+    > {
         // Check that account is forum member
         let who = Self::ensure_is_moderator(origin, &moderator_id)?;
 
         let thread = Self::ensure_thread_exists(category_id, thread_id)?;
 
-        Self::ensure_can_moderate_category(&who, moderator_id, thread.category_id)?;
+        Self::ensure_can_moderate_category(&who, moderator_id, &thread.category_id)?;
 
-        Ok(thread)
+        Ok((who, thread))
+    }
+
+    fn ensure_can_move_thread(
+        origin: T::Origin,
+        moderator_id: &T::ModeratorId,
+        category_id: &T::CategoryId,
+        thread_id: &T::ThreadId,
+        new_category_id: &T::CategoryId,
+    ) -> Result<
+        (
+            T::AccountId,
+            Thread<T::ForumUserId, T::CategoryId, T::Moment, T::Hash>,
+        ),
+        &'static str,
+    > {
+        ensure!(category_id != new_category_id, ERROR_THREAD_MOVE_INVALID,);
+
+        let info = Self::ensure_can_moderate_thread(origin, moderator_id, category_id, thread_id);
+        if info.is_err() {
+            return Err(ERROR_MODERATOR_MODERATE_ORIGIN_CATEGORY);
+        }
+        let (account_id, thread) = info.unwrap();
```

Comment:
This is not statically panic-free, but it can be, ill leave it for you to find the best way.